Source/WTF/wtf/glib/FileSystemGlib.cpp

```diff
@@ -185,30 +246,55 @@ PlatformFileHandle openFile(const String& path, FileOpenMode mode, FileAccessPer
         return invalidPlatformFileHandle;
 
     GRefPtr<GFile> file = adoptGRef(g_file_new_for_path(filename.data()));
-    GRefPtr<GFileIOStream> ioStream;
-    GFileCreateFlags permissionFlag = (permission == FileAccessPermission::All) ? G_FILE_CREATE_NONE : G_FILE_CREATE_PRIVATE;
-
-    if (failIfFileExists) {
-        ioStream = adoptGRef(g_file_create_readwrite(file.get(), permissionFlag, nullptr, nullptr));
-        return ioStream.leakRef();
+    GFileCreateFlags creationFlags = (permission == FileAccessPermission::All) ? G_FILE_CREATE_NONE : G_FILE_CREATE_PRIVATE;
+
+    if (mode == FileOpenMode::Read)
+        return G_SEEKABLE(g_file_read(file.get(), nullptr, nullptr));
+
+    // Embarrasingly, GIO doesn't have an atomic create-or-open function.
+    //
+    // The closest thing to create-or-open in GIO is g_file_replace() and g_file_replace_readwrite(), but they have some problems:
+    // (a) In the case of modifying an existing file, they return a new file in /tmp, doing a move swap when closing.
+    //     Even calling g_output_stream_flush() won't update the original file.
+    // (b) g_file_replace_readwrite() followed by a g_io_stream_close() will truncate the file, but we don't want ReadWrite to truncate.
+    //
+    // There is no pretty way out of this. Two naive solutions are:
+    // (a) We try to create, if it fails we try to open -> If the file is deleted between the two operations we fail.
+    // (b) We try to open, if it fails we try to create -> If the file is created between the two operations we fail.
+    //
+    // The most reliable thing we can do is spinlock retrying the other operation when we get a G_IO_ERROR_NOT_FOUND or G_IO_ERROR_EXISTS.
+    while (true) {
+        GRefPtr<GSeekable> seekable;
+        GError* error = nullptr;
+
+        // Try to create the file
+        if (mode == FileOpenMode::Truncate)
+            seekable = adoptGRef(G_SEEKABLE(g_file_create(file.get(), creationFlags, nullptr, &error)));
+        else
+            seekable = adoptGRef(G_SEEKABLE(g_file_create_readwrite(file.get(), creationFlags, nullptr, &error)));
+
+        if (seekable || failIfFileExists || (error && error->code != G_IO_ERROR_EXISTS))
+            return seekable.leakRef();
+
+        // Try to open the file instead.
+        // We need to open the file as readwrite because there is no write-only open in GIO.
+        // We cannot even extract a GFileOutputStream from the GFileIOStream after the fact: the closest thing is
+        // g_io_stream_get_output_stream(), but that returns a GOutputStream which is not a GFileOutputStream and
+        // doesn't implement the GSeekable interface.
```

Comment:
Did you try casting the GOutputStream* to a GFileOutputStream* to be sure it's not really a GFileOutputStream? It probably is. Would be really weird if not.

Reply:
Hm, you're right. I must have tested something wrong. It was pretty late. Changing that...

Reply:
nvm, I was getting the latter from a typo in an unrelated change.

Reply:
nvm²: "Stream is already closed" is emitted by any write attempt, so yeah, it's closed by the container GIOStream.